keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,641 @@
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [Inform how costly is to interrupt a Workload](#inform-how-costly-is-to-interrupt-a-workload)
+  - [Extra knobs in ClusterQueue preemption policy](#extra-knobs-in-clusterqueue-preemption-policy)
+<!-- /toc -->
+
+## Summary
+
+This enhancement introduces workload preemption, a mechanism to suspend
+workloads when:
+- ClusterQueues under their minimum quota need the resources that are currently
+  lent to other ClusterQueues in the cohort.
+- Within a Clusterqueue, there are running Workloads with lower priority than
+  a pending Workload.
+
+API fields in the Workload and ClusterQueue definitions determine preemption policies.
+
+## Motivation
+
+When ClusterQueues under their minimum quota lend resources, they should
+be able to recover those resources fast, to be able to admit Workloads
+when there are sudden spikes. Similarly, the ClusterQueue should be able to
+recover quota from low priority workloads that are currently running.
+
+Currently, the only mechanism to recover those resources is to wait for
+Workloads to finish, which is generally unbounded.
+
+### Goals
+
+- Preempt Workloads from ClusterQueues borrowing resources when other
+  ClusterQueues in the cohort, under their minimum quota, need the resources.
+- Preempt Workloads within a ClusterQueue when a high priority Workload doesn't
+  fit in the available quota, independently of borrowed quota.
+- Introduce API fields in ClusterQueue to control when preemption occurs.
+
+### Non-Goals
+
+- Graceful termination of Workloads is left to the workload pods to implement.
+- Tracking usage by workloads that take arbitrary time to be suspended. See
+  [Workload preemption doesn't imply immediate Pod termination](#workload-preemption-doesnt-imply-immediate-pod-termination) to learn more.
+  For example, the integration with Job uses the [suspend field](https://kubernetes.io/docs/concepts/workloads/controllers/job/#suspending-a-job).
+- Partial workload suspension is not supported.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+This enhancement proposes the introduction of the following APIs to influence
+Workload preemption:
+- A field in the ClusterQueue to influence how long to wait before starting
+  preemption.
+- A field in the Workload to determine whether to terminate or requeue upon preemption.
```

Comment:
What are the usecases in which you think termination will be useful ?

Reply:
I think I'd like a separation between workloads that went over their allocated timelimit and workloads that are over their timelimit but allowed to run due to preemption. 

I don't think it needs to be addressed here but that is why I created https://github.com/kubernetes-sigs/kueue/issues/405 to maybe discuss separately from preemption.